File: testData.py
import os.path

import numpy as np
import cv2
import torch
import json


def load_8bitImage(root):
    img = cv2.imread(root, -1)
    img = np.array(img, dtype=np.float32)
    img = torch.tensor(img).unsqueeze(2)
    img = np.array(img)
    img[np.isnan(img)] = 0

    return img.astype(np.float32)


def load_24bitNormal(root, R):
    nml = cv2.imread(root, -1)
    nml = np.array(nml, dtype=np.float32)
    nml = np.array(nml)
    nml[np.isnan(nml)] = 0
    nml = nml[:, :, [2, 1, 0]]
    mask = (nml[:, :, 0] == 0) & (nml[:, :, 1] == 0) & (nml[:, :, 2] == 0)

    nml[:, :, 0] = (~mask) * (nml[:, :, 0] / 255.0 * 2.0 - 1.0)
    nml[:, :, 1] = (~mask) * (nml[:, :, 1] / 255.0 * 2.0 - 1.0)
    nml[:, :, 2] = (~mask) * (1.0 - nml[:, :, 2] / 255.0 * 2.0)

    nml = R.transpose(0, 1) @ torch.tensor(nml).permute(2, 0, 1).reshape(3, -1)
    nml = nml.reshape(3, 512, 512).permute(1, 2, 0)

    return np.array(nml).astype(np.float32)


def load_scaled16bitImage(root, minVal, maxVal):
    img = cv2.imread(root, -1)
    img = np.array(img, dtype=np.float32)
    mask = (img == 0)
    img = img / 65535 * (maxVal - minVal) + minVal
    img[np.isnan(img)] = 0

    img = torch.tensor((~mask) * img).unsqueeze(2)
    img = np.array(img)

    return img.astype(np.float32)


def depth2vertex(depth, K, R, t):
    camOrig = -R.transpose(0, 1) @ t.unsqueeze(1)

    X = torch.arange(0, depth.size(2)).repeat(depth.size(1), 1) - K[0, 2]
    Y = torch.transpose(torch.arange(0, depth.size(1)).repeat(depth.size(2), 1), 0, 1) - K[1, 2]
    Z = torch.ones(depth.size(1), depth.size(2)) * K[0, 0]
    Dir = torch.cat((X.unsqueeze(2), Y.unsqueeze(2), Z.unsqueeze(2)), 2)

    vertex = Dir * (depth.squeeze(0) / torch.norm(Dir, dim=2)).unsqueeze(2).repeat(1, 1, 3)
    vertex = R.transpose(0, 1) @ vertex.permute(2, 0, 1).reshape(3, -1)
    vertex = camOrig.unsqueeze(1).repeat(1, 512, 512) + vertex.reshape(3, 512, 512)
    vertex = vertex.permute(1, 2, 0)
    return np.array(vertex)


def cameraVisualization():
    points = []
    for p in range(-50, 51):
        ps = float(p) / 100.0
        points.append([ps, 0.5, 0.5])
        points.append([ps, -0.5, 0.5])
        points.append([ps, 0.5, -0.5])
        points.append([ps, -0.5, -0.5])
        points.append([0.5, ps, 0.5])
        points.append([-0.5, ps, 0.5])
        points.append([0.5, ps, -0.5])
        points.append([-0.5, ps, -0.5])
        points.append([0.5, 0.5, ps])
        points.append([0.5, -0.5, ps])
        points.append([-0.5, 0.5, ps])
        points.append([-0.5, -0.5, ps])

    for p in range(-30, 31):
        ps = float(p) / 100.0
        points.append([ps, 0.3, 0.3 + 0.8])
        points.append([ps, -0.3, 0.3 + 0.8])
        points.append([ps, 0.3, -0.3 + 0.8])
        points.append([ps, -0.3, -0.3 + 0.8])
        points.append([0.3, ps, 0.3 + 0.8])
        points.append([-0.3, ps, 0.3 + 0.8])
        points.append([0.3, ps, -0.3 + 0.8])
        points.append([-0.3, ps, -0.3 + 0.8])
        points.append([0.3, 0.3, ps + 0.8])
        points.append([0.3, -0.3, ps + 0.8])
        points.append([-0.3, 0.3, ps + 0.8])
        points.append([-0.3, -0.3, ps + 0.8])

    return points


def lightVisualization():
    points = []
    for px in range(-5, 6):
        for py in range(-5, 6):
            for pz in range(-5, 6):
                p = np.array([px, py, pz]).astype(np.float) / 10
                if np.linalg.norm(p) > 0:
                    points.append(p / np.linalg.norm(p))
    return points


def writePLY(vertex, normal, image, mask, filename, cameraPoints=None, lightPoints=None):
    numPoints = np.sum(mask)
    if cameraPoints is not None:
        numPoints += len(cameraPoints)
    if lightPoints is not None:
        numPoints += len(lightPoints)

    ply_file = open(filename, "w")
    ply_file.write('ply\n')
    ply_file.write('format ascii 1.0\n')
    ply_file.write('element vertex ' + str(numPoints) + '\n')
    ply_file.write('property float x\n')
    ply_file.write('property float y\n')
    ply_file.write('property float z\n')
    ply_file.write('property float nx\n')
    ply_file.write('property float ny\n')
    ply_file.write('property float nz\n')
    ply_file.write('property uchar red\n')
    ply_file.write('property uchar green\n')
    ply_file.write('property uchar blue\n')
    ply_file.write('end_header\n')

    for i in range(vertex.shape[0]):
        for j in range(vertex.shape[1]):
            if mask[i, j] is True:
                ply_file.write(f"{' '.join(map(str, vertex[i, j, :]))} "
                               f"{' '.join(map(str, normal[i, j, :]))} "
                               f"{' '.join(map(str, image[i, j, :].repeat(3).astype(np.int32)))}\n")
    if cameraPoints is not None:
        for i in range(len(cameraPoints)):
            ply_file.write(f"{' '.join(map(str, cameraPoints[i]))} 0.0 0.0 0.0 255 0 0\n")
    if lightPoints is not None:
        for i in range(len(lightPoints)):
            ply_file.write(f"{' '.join(map(str, lightPoints[i]))} 0.0 0.0 0.0 255 255 0\n")

    ply_file.close()


# ----------------------------------------------------------------------------------
path = 'SyntheticDataSet/CapturedData/0000'
# path = 'RealDataSet/0000.'

file_idx = 0
file = path + str(file_idx) + ".data0" + ".json"

while os.path.isfile(file):
    f = open(file)
    data = json.load(f)
    f.close()

    image = load_8bitImage(path + str(file_idx) + f'.image0.png')
    depth = load_scaled16bitImage(path + str(file_idx) + f'.depth0.png', data['minDepth'], data['maxDepth'])
    vertex = depth2vertex(torch.tensor(depth).permute(2, 0, 1), torch.tensor(data['K']), torch.tensor(data['R']),
                          torch.tensor(data['t']))
    normal = load_24bitNormal(path + str(file_idx) + f'.normal0.png', torch.tensor(data['R']))
    # mask = np.sum(np.abs(vertex), axis=2) != 0
    mask = np.sum(np.abs(depth), axis=2) != 0
    cameraPoints = cameraVisualization()
    for i in range(len(cameraPoints)):
        cameraPoints[i] = np.array(data['R']).transpose() @ cameraPoints[i] / 4 - np.array(
            data['R']).transpose() @ np.array(data['t'])
    lightPoints = lightVisualization()
    for i in range(len(lightPoints)):
        lightPoints[i] = lightPoints[i] / 8 + np.array(data['lightPos'])

    writePLY(vertex, normal, image, mask, path + str(file_idx) + f'.pointCloud0.ply',
             cameraPoints=cameraPoints,
             lightPoints=lightPoints)

    file_idx += 1
    file = path + str(file_idx) + ".data0" + ".json"

File: help_funs/testData.m
path = 'SyntheticDataSet\CapturedData\00000.';
%path = 'RealDataSet\00000.';

%load image
C.image = imread([path, 'image0.png']);
C.image = cat(3, C.image, C.image, C.image);

%load data, containing calibration, min/max depths
fid = fopen([path, 'data0.json'], 'r'); 
str = char(fread(fid,inf)'); 
fclose(fid); 
C.data = jsondecode(str);

%load depth map
C.depth = double(imread([path, 'depth0.png']));
C.mask = C.depth ~= 0;
C.depth = C.mask .* (C.depth*(C.data.maxDepth-C.data.minDepth)/65535 + C.data.minDepth);

%compute vertex map from depth map and calibration
C.vertex = depth2vertex(C.depth, C.data.K);

%load normals in tangent space and transform to normal space, 
%apply rotation, if camera is rotated
normal_tan = double(imread([path, 'normal0.png']));
normalOrig = cat(3, C.mask, C.mask, -C.mask) .* (cat(3, normal_tan(:,:,1), normal_tan(:,:,2), normal_tan(:,:,3))/255*2 - 1);
C.normal = normalOrig;
C.normal(:,:,1) = C.data.R(1,1)*normalOrig(:,:,1) + C.data.R(2,1)*normalOrig(:,:,2) + C.data.R(3,1)*normalOrig(:,:,3);
C.normal(:,:,2) = C.data.R(1,2)*normalOrig(:,:,1) + C.data.R(2,2)*normalOrig(:,:,2) + C.data.R(3,2)*normalOrig(:,:,3);
C.normal(:,:,3) = C.data.R(1,3)*normalOrig(:,:,1) + C.data.R(2,3)*normalOrig(:,:,2) + C.data.R(3,3)*normalOrig(:,:,3);

%compute point clouds and visualizations of the camera and the light source
C.pointCloud = RGBDN2pointCloud(C.image, C.depth, C.normal, C.data);

camera = camera_visualization/4;
light = light_visualization/8;
camera = C.data.R'*camera - repmat(C.data.R'*C.data.t,1,size(camera,2));
light = light + repmat(C.data.lightPos,1,size(light,2));
camera = [camera; zeros(size(camera)); 255*ones(size(camera))]; 
camera(8:9,:) = 0;
light = [light; zeros(size(light)); 255*ones(size(light))];
light(9,:) = 0;

savePLY([path, 'pointcloud0.ply'], [C.pointCloud, camera, light]);

function vertex = depth2vertex(depth, K)

    [X,Y] = meshgrid(1:size(depth,2), 1:size(depth,1));
    X = X - K(1,3);
    Y = Y - K(2,3);
    Z = ones(size(depth)) * K(1,1);
    Dir = cat(3,X,Y,Z);
    vertex = Dir .* repmat(depth ./ sqrt(sum(Dir.^2,3)),1,1,3);
end

function pointCloud = RGBDN2pointCloud(im, depth, normal, P)

    camOrig = -P.R' * P.t;
    %triangulate point cloud from cameras and matches
    N = sum(sum(depth ~= 0));
    pointCloud = zeros(9, N);
    n = 1;
    for x = 1 : size(depth, 2)
        for y = 1 : size(depth, 1)
            if depth(y, x) ~= 0
                dir = [x - P.K(1,3); y - P.K(2,3); P.K(1,1)];
                pointCloud(1:3,n) = camOrig + P.R' * dir * depth(y, x) / norm(dir);
                pointCloud(4:6,n) = squeeze(normal(y,x,:));
                pointCloud(7:9,n) = squeeze(im(y,x,:));
                n = n+1;
            end
        end
    end
end

function savePLY( dest, points )

    N = size(points, 2);
    fileID = fopen(dest, 'w');
    fprintf(fileID, 'ply\n');
    fprintf(fileID, 'format binary_little_endian 1.0\n');
    fprintf(fileID, ['element vertex ', num2str(N), '\n']);
    fprintf(fileID, 'property float x\n');
    fprintf(fileID, 'property float y\n');
    fprintf(fileID, 'property float z\n');
    fprintf(fileID, 'property float nx\n');
    fprintf(fileID, 'property float ny\n');
    fprintf(fileID, 'property float nz\n');
    fprintf(fileID, 'property uchar red\n');
    fprintf(fileID, 'property uchar green\n');
    fprintf(fileID, 'property uchar blue\n');
    fprintf(fileID, 'end_header\n');

    colors = 255 * ones(3, N);
    normals = zeros(3, N);
    if size(points, 1) == 6
        colors = points(4:6,:);
    end
    if size(points, 1) == 9
        normals = points(4:6,:);
        colors = points(7:9,:);
    end

    points_8bit = reshape(typecast(single(reshape(points(1:3,:), 3*N, 1)), 'uint8'), 12, N);
    normals_8bit = reshape(typecast(single(reshape(normals, 3*N, 1)), 'uint8'), 12, N);
    combined = reshape([points_8bit; normals_8bit; colors], 27*N, 1);

    fwrite(fileID, combined, 'uint8');
    fclose(fileID);
end

function C = loadData(path, i, net)
     
    %compute point clouds
    camera = camera_visualization/4;
    light = light_visualization/8;
    camera = C.data.R'*camera - repmat(C.data.R'*C.data.t,1,size(camera,2));
    light = light + repmat(C.data.lightPos,1,size(light,2));
    camera = [camera; zeros(size(camera)); 255*ones(size(camera))]; 
    camera(8:9,:) = 0;
    light = [light; zeros(size(light)); 255*ones(size(light))];
    light(9,:) = 0;
    
    C.pointCloud = RGBDN2pointCloud(C.image, C.depth, C.normal, C.data);
    savePLY([path, 'pointcloud', num2str(i), net, '.ply'], [C.pointCloud, camera, light]);
end

function points = camera_visualization(  )

points = [];

for px = -0.5 : 0.01 : 0.5
    for py = -0.5 : 0.01 : 0.5
        for pz = -0.5 : 0.01 : 0.5
            p = [px; py; pz];
            p_sorted = sort(abs(p));
            if p_sorted(2) == 0.5 && p_sorted(3) == 0.5
                points = [points, p];
            end
        end
    end
end
for px = -0.3 : 0.01 : 0.3
    for py = -0.3 : 0.01 : 0.3
        for pz = -0.3 : 0.01 : 0.3
            p = [px; py; pz];
            p_sorted = sort(abs(p));
            if p_sorted(2) == 0.3 && p_sorted(3) == 0.3
                p = [px; py; pz+0.8];
                points = [points, p];
            end
        end
    end
end

end

function points = projector_visualization(  )

points = [];

for px = -0.5 : 0.01 : 0.8
    for py = -0.5 : 0.01 : 0.5
        for pz = -0.5 : 0.01 : 0.5
            p = [px; py; pz];
            p_sorted = sort(abs(p));
            if p_sorted(2) == 0.5 && p_sorted(3) == 0.5
                points = [points, p];
            end
        end
    end
end
points(2,:) = points(2,:)/2;
points(1,:) = points(1,:)+0.25;
points(3,:) = points(3,:)*0.8 + 0.2;

for px = -0.2 : 0.01 : 0.2
    for py = -0.2 : 0.01 : 0.2
        for pz = -0.2 : 0.01 : 0.2
            p = [px; py; pz];
            p_sorted = sort(abs(p));
            if p_sorted(2) == 0.2 && p_sorted(3) == 0.2
                p = [px; py; pz+0.8];
                points = [points, p];
            end
        end
    end
end

end

function points = light_visualization(  )

points = [];

for px = -0.5 : 0.1 : 0.5
    for py = -0.5 : 0.1 : 0.5
        for pz = -0.5 : 0.1 : 0.5
            p = [px; py; pz];
            if norm(p) > 0
                points = [points, p/norm(p)];
            end
        end
    end
end

end